react/CreditCard.tsx

```diff
@@ -165,101 +162,156 @@ const CreditCard: React.FC<Props> = ({
-        showError: true,
-        errorMessage: intl.formatMessage(messages.invalidDigits),
-      })
-      return false
-    }
-
-    return true
-  }
+  const [submitLoading, setSubmitLoading] = useState(false)
 
   const handleSubmit = async () => {
-    const docIsValid = validateDoc()
-    const { data: cardIsValid } = await postRobot.send(
-      iframeRef.current!.contentWindow,
-      'isCardValid'
-    )
-
-    if (
-      !selectedPaymentSystem ||
-      !cardIsValid ||
-      (cardType === 'new' && !docIsValid)
-    ) {
-      showCardErrors()
-      return
-    }
+    setSubmitLoading(true)
 
-    if (cardType === 'new') {
-      const { data: lastDigits } = await postRobot.send(
+    try {
+      const { data: cardIsValid } = await postRobot.send(
         iframeRef.current!.contentWindow,
-        'getCardLastDigits'
+        'isCardValid'
       )
 
-      setCardLastDigits(lastDigits)
+      if (!selectedPaymentSystem || !cardIsValid) {
+        showCardErrors()
+        return
+      }
 
-      await setPaymentField({
-        paymentSystem: selectedPaymentSystem.id,
-        referenceValue,
-        installments: null,
-      })
-    }
+      if (cardType === 'new') {
+        const { data: lastDigits } = await postRobot.send(
+          iframeRef.current!.contentWindow,
+          'getCardLastDigits'
+        )
 
-    onCardFormCompleted()
-  }
+        setCardLastDigits(lastDigits)
 
-  const handleChangeDoc = (data: any) => {
-    setDoc({
-      value: data.document,
-      error: !data.isValid,
-      errorMessage: '',
-      showError: false,
-    })
+        await setPaymentField({
+          paymentSystem: selectedPaymentSystem.id,
+          referenceValue,
+          installments: null,
+        })
+      }
+
+      onCardFormCompleted()
+    } finally {
+      setSubmitLoading(false)
+    }
   }
 
   const handleCardSummaryClick = async () => {
     resetCardFormData()
     onChangePaymentMethod()
   }
 
+  const [
+    showAvailableInstallmentOptionsModal,
+    setShowAvailableInstallmentOptionsModal,
+  ] = useState(false)
+  const { orderForm } = useOrderForm()
+
+  const creditCardPayments = useMemo(
+    () =>
+      orderForm.paymentData.paymentSystems.filter(
+        ({ groupName }) => groupName === 'creditCardPaymentGroup'
+      ),
+    [orderForm]
+  )
+
+  const creditCardInstallmentOptions = useMemo(
+    () =>
+      creditCardPayments
+        .map(creditCardPaymentSystem => {
+          const creditCardInstallments = orderForm.paymentData.installmentOptions.find(
+            installmentOption =>
+              creditCardPaymentSystem.id === installmentOption.paymentSystem
+          )
+
+          if (!creditCardInstallments) {
+            return null
+          }
+
+          return {
+            ...creditCardInstallments,
+            paymentName: creditCardPaymentSystem.name,
+          }
+        })
+        .filter(<T extends any>(value: T | null): value is T => value !== null),
```

Comment:
question (non-blocking): what that does?

Reply:
removes `null` from the return type, so typescript can infer that the array doesn't contain null values